File: LeetCode/力扣杯/LCP 31. 变换的迷宫.cpp
bool f[110][51][51][2][3];

class Solution {
private:
    static constexpr int dirs[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    
public:
    bool escapeMaze(vector<vector<string>>& maze) {
        int m = maze[0].size();
        int n = maze[0][0].size();
        if (m == 1 && n == 1) {
            return true;
        }
        
        int k = maze.size();
        memset(f, false, sizeof(f));
        f[0][0][0][0][0] = true;
        
        for (int r = 0; r < k - 1; ++r) {
            for (int i = 0; i < m; ++i) {
                for (int j = 0; j < n; ++j) {
                    for (int u1 = 0; u1 < 2; ++u1) {
                        for (int v1 = 0; v1 < 3; ++v1) {
                            if (!f[r][i][j][u1][v1]) {
                                continue;
                            }
                            // move
                            for (int d = 0; d < 4; ++d) {
                                int ni = i + dirs[d][0];
                                int nj = j + dirs[d][1];
                                if (ni >= 0 && ni < m && nj >= 0 && nj < n) {
                                    if (maze[r + 1][ni][nj] == '.') {
                                        f[r + 1][ni][nj][u1][v1 ? 2 : 0] = true;
                                    }
                                    else {
                                        if (!u1)
                                            f[r + 1][ni][nj][1][v1 ? 2 : 0] = true;
                                        if (!v1)
                                            f[r + 1][ni][nj][u1][1] = true;
                                    }
                                }
                            }
                            // stand
                            if (maze[r + 1][i][j] == '.') {
                                f[r + 1][i][j][u1][v1] = true;
                            }
                            else {
                                if (v1 == 1)
                                    f[r + 1][i][j][u1][v1] = true;
                                if (!u1) 
                                    f[r + 1][i][j][1][v1] = true;
                                if (!v1)
                                    f[r + 1][i][j][u1][1] = true;
                                
                            }
                        }
                    }
                }
            }

        }
        
        for (int r = 0; r < k; ++r) {
            for (int u1 = 0; u1 < 2; ++u1) {
                for (int v1 = 0; v1 < 3; ++v1) {
                    if (f[r][m - 1][n - 1][u1][v1]) {
                        return true;
                    }
                }
            }
        }
        return false;
    }
};

File: LeetCode/力扣杯/LCP 31. 变换的迷宫.py

class Solution:
    def escapeMaze(self, maze: List[List[str]]) -> bool:
        q = deque()
        T = len(maze)
        n, m = len(maze[0]), len(maze[0][0])
        # visited = set()
        t = 0
        for i in range(n):
            for j in range(m):
                q.append((0,0,0,0))
        while q:
            sz = len(q)
            visited = set()
            for _ in range(sz):
                row, col, tmp, per = q.popleft()
                if (row,col) == (n-1,m-1):
                    return True
                if n-1-row+m-1-col > T-1-t:
                    continue
                if t+1 < T:
                    for dr, dc in [(0,0),(0,1),(0,-1),(1,0),(-1,0)]:
                        dr += row
                        dc += col
                        if dr < 0 or dr >= n or dc < 0 or dc >= m:
                            continue
                        if (dr,dc) != (row,col) and per == 1:
                            per = 2
                        if maze[t+1][dr][dc] == ".":
                            # if (dr,dc) != (row,col) and per == 1:
                            #     per = 2
                            if (dr,dc,tmp,per) not in visited:
                                visited.add((dr,dc,tmp,per))
                                q.append((dr,dc,tmp,per))
                        else:
                            if per == 1 and (dr,dc,tmp,per) not in visited:
                                visited.add((dr,dc,tmp,per))
                                q.append((dr,dc,tmp,per))
                            if tmp == 0 and (dr,dc,1,per) not in visited:
                                visited.add((dr,dc,1,per))
                                q.append((dr,dc,1,per))
                            if per == 0 and (dr,dc,tmp,1) not in visited:
                                visited.add((dr,dc,tmp,1))
                                q.append((dr,dc,tmp,1))

                            
            t += 1
 
        return False